### ISSUE TITLE
[Query] Cluster operations / delay in sync

### ISSUE THREAD

**Issue by zehawki** [typesense/typesense#487]
1. Can you please mention some details of how clustering works, esp how quickly do the nodes sync up when multi region? Are sync done based on time (interval), or based on write events? What happens if nodes have different values, is that then settled based on last writer?
2. How stable is Typesense clustering for scenarios of continuous updates? Sort of related to this https://github.com/typesense/typesense/issues/450. Are there any gotchas here to be aware of?

**Comment by kishorenc** [comment]
1. Clustering is implemented using the [Raft consensus algorithm](https://raft.github.io/). In Raft, each cluster has an elected leader which handles all writes. Every write must be acknowledge by a majority of node before it is deemed a success (2 nodes in a 3 node cluster). Since all writes are handled by an elected reader, there are no conflicts. If a network partition occurs, the majority partition will elect a new leader and continue writing, while the minority partition will reject the writes. On rejoining all nodes will re-sync data from the current leader so everything is consistent. Typesense reads can lag writes by a second (depending on the geographical distance) or so since we have chosen availability over consistency. 

2. Continuous updates are fine, but batching them up (2000-5000 records per batch) is wise in high write volume scenarios, since the writes are synchronous. As a search engine, we need to protect the searches so if you try to write too fast, in order to preserve search latency, we can reject writes with a 503 status code with  a `Not ready or lagging` error message. The client should detect this and back off appropriately.

**Comment by orasik** [comment]
@kishorenc thank you for the detailed answer.
I have a question about lagging. So I was testing taking nodes down and bringing them back again to test if they will index the data from the disk again which they did.

However, for a large collection (1M+), I can't search the collection unless all nodes had finished indexing again.

Is there a way that I can change this behavior to search in the already loaded documents? For instance, let's assume that after 10 minutes, I had 100K documents loaded successfully. I want to be able to search this 100 K instead of waiting for the other 900K to load.

**Comment by kishorenc** [comment]
> I can't search the collection unless all nodes had finished indexing again.

This should not happen in a HA set-up. When a node goes down, the other nodes are not affected by that. Are you talking about a single node mode?

>  I want to be able to search this 100 K instead of waiting for the other 900K to load.

This is not possible at the moment because, from past experience, allowing such partial searches has proven to be very confusing both for users and operationally. In any case allowing searches during the initial restoring phase will involve high search latencies as the searches will compete with the bulk loading/indexing of documents.

I recommend using a 3-node HA configuration to work around this issue. Our client libraries accept multiple nodes in the configuration and can work around a bad node by omitting it from the pool.

**Comment by orasik** [comment]
> This should not happen in a HA set-up. When a node goes down, the other nodes are not affected by that. Are you talking about a single node mode?

No, I was talking about HA. One of the tests I've done was taking all the three nodes down to see how long it will take to restore the collection. I know it is an extreme case but just wanted to be prepared.
